daemon/api_sideload_n_try_test.go

```diff
@@ -645,6 +645,128 @@ func (s *sideloadSuite) TestCleanUpUnusedTempSnapFiles(c *check.C) {
+		"\r\n" +
+		"true\r\n" +
+		"----hello--\r\n"
+	for _, snap := range snaps {
+		body += "Content-Disposition: form-data; name=\"snap\"; filename=\"file-" + snap + "\"\r\n" +
+			"\r\n" +
+			snap + "\r\n" +
+			"----hello--\r\n"
+	}
+
+	req, err := http.NewRequest("POST", "/v2/snaps", bytes.NewBufferString(body))
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/thing; boundary=--hello--")
+	s.asUserAuth(c, req)
+	rsp := s.asyncReq(c, req, s.authUser)
+
+	st := d.Overlord().State()
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.Change(rsp.Change)
+	c.Assert(chg, check.NotNil)
+	c.Check(chg.Summary(), check.Equals, fmt.Sprintf(`Install snaps %q from files ["file-one" "file-two"]`, snaps))
+
+	var data map[string][]string
+	c.Assert(chg.Get("api-data", &data), check.IsNil)
+	c.Check(data["snap-names"], testutil.DeepUnsortedMatches, snaps)
+}
+
+func (s *sideloadSuite) TestSideloadManyFailInstallPathMany(c *check.C) {
+	s.daemon(c)
+	restore := daemon.MockSnapstateInstallPathMany(func(_ context.Context, s *state.State, infos []*snap.SideInfo, paths []string, userID int, flags *snapstate.Flags) ([]*state.TaskSet, error) {
+		return nil, errors.New("expected")
+	})
+	defer restore()
+
+	readRest := daemon.MockUnsafeReadSnapInfo(func(string) (*snap.Info, error) {
+		return &snap.Info{SuggestedName: "name"}, nil
+	})
+	defer readRest()
+
+	body := "----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"devmode\"\r\n" +
+		"\r\n" +
+		"true\r\n" +
+		"----hello--\r\n"
+	for _, snap := range []string{"one", "two"} {
+		body += "Content-Disposition: form-data; name=\"snap\"; filename=\"file-" + snap + "\"\r\n" +
+			"\r\n" +
+			"xyzzy \r\n" +
+			"----hello--\r\n"
+	}
+
+	req, err := http.NewRequest("POST", "/v2/snaps", bytes.NewBufferString(body))
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/thing; boundary=--hello--")
+	apiErr := s.errorReq(c, req, nil)
+
+	c.Check(apiErr.JSON().Status, check.Equals, 500)
+	c.Check(apiErr.Message, check.Equals, `cannot install snap files: expected`)
+}
+
+func (s *sideloadSuite) TestSideloadManyFailUnsafeReadInfo(c *check.C) {
+	s.daemon(c)
+	restore := daemon.MockUnsafeReadSnapInfo(func(string) (*snap.Info, error) {
+		return nil, errors.New("expected")
+	})
+	defer restore()
+
+	body := "----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"devmode\"\r\n" +
+		"\r\n" +
+		"true\r\n" +
+		"----hello--\r\n"
+	for _, snap := range []string{"one", "two"} {
+		body += "Content-Disposition: form-data; name=\"snap\"; filename=\"file-" + snap + "\"\r\n" +
+			"\r\n" +
+			"xyzzy \r\n" +
+			"----hello--\r\n"
+	}
+
+	req, err := http.NewRequest("POST", "/v2/snaps", bytes.NewBufferString(body))
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/thing; boundary=--hello--")
+	apiErr := s.errorReq(c, req, nil)
+
+	c.Check(apiErr.JSON().Status, check.Equals, 400)
+	c.Check(apiErr.Message, check.Equals, `cannot read snap file: expected`)
+}
+
```

Comment:
We probably need another test where dangerous or devmode are not set and one of the snaps isn't asserted.